src/coreclr/jit/valuenum.cpp

```diff
@@ -3816,14 +3832,18 @@ ValueNum ValueNumStore::VNApplySelectorsAssignTypeCoerce(ValueNum elem, var_type
                 // We are trying to write an 'elem' of type 'elemType' using 'indType' store
 
                 // insert a cast of elem to 'indType'
-                elem = VNForCast(elem, indType, elemTyp);
+                dstVN = VNForCast(srcVN, dstIndType, srcType);
```

Comment:
Maybe assert we only have integral types here? I always wondered if it is possible to construct a case where VN will insert a cast, say `float -> double` (or `float` -> `int`, etc), on mismatched assignment, where it was actually meant to be a `BITCAST`. Looks like like morph is careful enough to not allow that to happen, but then maybe I am just not imaginative enough.

Reply:
such assert won't stand nowadays, first because there are many cases where `dstIndType` is not integral, second because VN tries to handle GC heap and for it the type is always `TYP_REF`, so we could have `srcType = TYP_REF`, `dstIndType = TYP_DOUBLE`.

There are many cases where VN is potentially wrong and we uncover more of them by improving our optimizations but for now, the CQ benefits from it still outweigh potential undiscovered bugs from it.

Reply:
I see, thank you for the explanation.